src/core/util/hugepage_mgr.h

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2001-2023 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef _XLIO_UTIL_HUGEPAGE_MGR_H_
+#define _XLIO_UTIL_HUGEPAGE_MGR_H_
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include <string>
+#include <vector>
+#include <unordered_map>
+
+#include "utils/lock_wrapper.h"
+
+typedef struct {
+    uint32_t nr_hugepages_total;
+    uint32_t nr_hugepages_free;
+    unsigned nr_hugepages_allocated;
+    unsigned nr_allocations;
+} hugepage_metadata;
+
+class hugepage_mgr {
+public:
+    hugepage_mgr();
+
+    void update();
+
+    inline size_t get_default_hugepage() { return m_default_hugepage; }
+
+    void *alloc_hugepages(size_t &size);
+    void dealloc_hugepages(void *ptr, size_t size);
+
+    void print_report(bool short_report = false);
+
+private:
+    enum {
+        HUGEPAGE_METRIC_OPTIMAL = 10U,
+        HUGEPAGE_METRIC_ACCEPTABLE = 50U,
+        HUGEPAGE_UNUSED_OPTIMAL = (2U * 1024U * 1024U),
+        HUGEPAGE_UNUSED_ACCEPTABLE = (256U * 1024U * 1024U),
+        HUGEPAGE_UNUSED_WARNING_THRESHOLD = (100U * 1024U * 1024U),
+    };
+
+    void read_sysfs();
+    size_t read_meminfo(const char *tag);
+    uint32_t read_file_uint(std::string path);
+
+    bool is_hugepage_optimal(size_t hugepage, size_t size);
+    bool is_hugepage_acceptable(size_t hugepage, size_t size);
+    size_t find_optimal_hugepage(size_t size);
+
+    inline size_t hugepage_unused_space(size_t hugepage, size_t size)
+    {
+        const size_t hugepage_mask = hugepage - 1;
```

Comment:
Can you please add a comment that explains what this method does (why not just hugepage - size).

Reply:
done.

`size` is the allocation request, it can be multiple hugepages. And the method returns unused space in the last hugepage because of alignment.